packages/compass-connections/src/connection-scoped-global-app-registry.ts

```diff
@@ -0,0 +1,66 @@
+import type { ConnectionInfo } from '@mongodb-js/connection-info';
+import {
+  type AppRegistry,
+  createServiceLocator,
+  useGlobalAppRegistry,
+} from 'hadron-app-registry';
+import { useConnectionInfo } from '@mongodb-js/connection-storage/provider';
+import { useMemo } from 'react';
+
+type ConnectionScopedGlobalAppRegistryEvents =
+  | 'schema-analyzed'
+  | 'documents-refreshed'
+  | 'documents-paginated'
+  | 'document-inserted';
+
+export type ConnectionScopedGlobalAppRegistryLocator<
+  T extends ConnectionScopedGlobalAppRegistryEvents,
+  K extends keyof ConnectionScopedGlobalAppRegistryImpl<T> = keyof ConnectionScopedGlobalAppRegistryImpl<T>,
+  L extends keyof ConnectionScopedGlobalAppRegistryImpl<T> = K
+> = () => ConnectionScopedGlobalAppRegistry<T, K, L>;
+
+export type ConnectionScopedGlobalAppRegistry<
+  T extends ConnectionScopedGlobalAppRegistryEvents,
+  K extends keyof ConnectionScopedGlobalAppRegistryImpl<T> = keyof ConnectionScopedGlobalAppRegistryImpl<T>,
+  L extends keyof ConnectionScopedGlobalAppRegistryImpl<T> = K
+> = Pick<ConnectionScopedGlobalAppRegistryImpl<T>, K> &
+  Partial<Pick<ConnectionScopedGlobalAppRegistryImpl<T>, L>>;
+
+interface EventForwarder<T extends ConnectionScopedGlobalAppRegistryEvents> {
+  emit(event: T, payload?: Record<string, any>): void;
+}
+
+export class ConnectionScopedGlobalAppRegistryImpl<
+  T extends ConnectionScopedGlobalAppRegistryEvents
+> implements EventForwarder<T>
+{
+  constructor(
+    private readonly appRegistryEmitter: AppRegistry['emit'],
+    private readonly sourceConnectionInfo: ConnectionInfo
+  ) {}
+
+  emit(event: T, payload: Record<string, any> = {}): void {
+    payload.sourceConnectionInfo = this.sourceConnectionInfo;
+    this.appRegistryEmitter(event, payload);
+  }
+}
+
+export const connectionScopedGlobalAppRegistryLocator = createServiceLocator(
+  function useConnectionScopedGlobalAppRegistry<
+    T extends ConnectionScopedGlobalAppRegistryEvents,
+    K extends keyof ConnectionScopedGlobalAppRegistryImpl<T> = keyof ConnectionScopedGlobalAppRegistryImpl<T>,
+    L extends keyof ConnectionScopedGlobalAppRegistryImpl<T> = K
+  >(): ConnectionScopedGlobalAppRegistry<T, K, L> {
+    const appRegistry = useGlobalAppRegistry();
+    const connectionInfo = useConnectionInfo();
```

Comment:
Once again: do we need this to be aware of all changes to the connection info and updated every time they happen or just getting an id here is fine to be passed with the payload? If, for some reason we want the full thing, can this be resolved when sending the event instead of subscribing the hook here and causing unnecessary updates?